packages/actor-query-operation-leftjoin-nestedloop/lib/ActorQueryOperationLeftJoinNestedLoop.ts

```diff
@@ -20,37 +24,71 @@ export class ActorQueryOperationLeftJoinNestedLoop extends ActorQueryOperationTy
   public async runOperation(pattern: Algebra.LeftJoin, context: ActionContext)
     : Promise<IActorQueryOperationOutputBindings> {
 
-    // uses nested loop join
-    const left: IActorQueryOperationOutputBindings = ActorQueryOperation.getSafeBindings(
-      await this.mediatorQueryOperation.mediate({ operation: pattern.left, context }));
-    const right: IActorQueryOperationOutputBindings = ActorQueryOperation.getSafeBindings(
-      await this.mediatorQueryOperation.mediate({ operation: pattern.right, context }));
-    const bindingsStream = left.bindingsStream.transform<Bindings>({
-      optional: true,
-      transform: (leftItem, next) => {
-        const rightStream = right.bindingsStream.clone();
-        rightStream.on('end', next);
-        rightStream.on('data', (rightItem) => {
-          const join = ActorRdfJoin.join(leftItem, rightItem);
-          if (join) {
-            // if (pattern.expression) {
-            //   // TODO: do this once expressions are implemented
-            //   // Values of both streams can be used for this expression
-            // }
-
-            bindingsStream._push(join);
-          }
+    const leftRaw = await this.mediatorQueryOperation.mediate({ operation: pattern.left, context });
+    const left = ActorQueryOperation.getSafeBindings(leftRaw);
+    const rightRaw = await this.mediatorQueryOperation.mediate({ operation: pattern.right, context });
+    const right = ActorQueryOperation.getSafeBindings(rightRaw);
+
+    const leftJoinInner = (outerItem: Bindings, innerStream: ClonedIterator<Bindings>) => {
+      const joinedStream = innerStream
+        .transform({
+          transform: async (innerItem: Bindings, nextInner: any) => {
+            const joinedBindings = ActorRdfJoin.join(outerItem, innerItem);
+            if (!joinedBindings) { nextInner(); return; }
+            if (!pattern.expression) {
+              joinedStream._push({ joinedBindings, result: true });
+              nextInner();
+              return;
+            }
+            try {
+              const evaluator = new AsyncEvaluator(pattern.expression);
```

Comment:
Can you move this evaluator outside of this transformer function? I assume it can be reused for all invocations?